monkey/agent_plugins/exploiters/mssql/src/plugin.py

```diff
@@ -0,0 +1,128 @@
+
+from .mssql_command_builder import build_mssql_command
+from .mssql_options import MSSQLOptions
+from .mssql_remote_access_client_factory import MSSQLRemoteAccessClientFactory
+
+logger = logging.getLogger(__name__)
+
+
+def should_attempt_exploit(host: TargetHost, target_ports: Sequence[NetworkPort]) -> bool:
+    relevant_tcp_ports_on_host = get_open_tcp_ports(
+        host=host, services_to_filter={NetworkService.MSSQL, NetworkService.UNKNOWN}
+    )
+    open_target_port_on_host_exists = any(
+        [port in relevant_tcp_ports_on_host for port in target_ports]
+    )
+
+    return open_target_port_on_host_exists or any_tcp_port_status_is_unknown(host, target_ports)
+
+
+class Plugin:
+    def __init__(
+        self,
+        *,
+        plugin_name: str,
+        agent_id: AgentID,
+        agent_event_publisher: IAgentEventPublisher,
+        agent_binary_repository: IAgentBinaryRepository,
+        propagation_credentials_repository: IPropagationCredentialsRepository,
+        otp_provider: IAgentOTPProvider,
+        **kwargs,
+    ):
+        self._plugin_name = plugin_name
+        self._agent_id = agent_id
+        self._agent_event_publisher = agent_event_publisher
+        self._agent_binary_repository = agent_binary_repository
+        credentials_generator = partial(
+            generate_brute_force_credentials,
+            identity_filter=identity_type_filter([Username]),
+            secret_filter=secret_type_filter([Password]),
+        )
+        self._credentials_provider = BruteForceCredentialsProvider(
+            propagation_credentials_repository, credentials_generator
+        )
+        self._otp_provider = otp_provider
+
+    def run(
+        self,
+        *,
+        host: TargetHost,
+        servers: Sequence[str],
+        current_depth: int,
+        options: Dict[str, Any],
+        interrupt: Event,
+        **kwargs,
+    ) -> ExploiterResultData:
+        # HTTP ports options are hack because they are needed in fingerprinters
+        del_key(options, "http_ports")
+
+        try:
+            logger.debug(f"Parsing options: {pformat(options)}")
+            mssql_options = MSSQLOptions(**options)
+        except Exception as err:
+            msg = f"Failed to parse MSSQL options: {err}"
+            logger.exception(msg)
+            return ExploiterResultData(error_message=msg)
+
+        if not should_attempt_exploit(host, mssql_options.target_ports):
+            msg = f"Host {host.ip} has no open MSSQL ports"
+            logger.debug(msg)
+            return ExploiterResultData(
+                exploitation_success=False, propagation_success=False, error_message=msg
+            )
+
+        command_builder = partial(
+            build_mssql_command,
+            self._agent_id,
+            servers,
+            current_depth,
+            remote_agent_binary_destination_path=get_agent_dst_path(host),
+            otp_provider=self._otp_provider,
+        )
```

Comment:
MSSQL has two commands and hence, no `build_mssql_command()`.

Reply:
MSSQL seems like a brute-force exploiter but uses HTTP server to download the agent. We may need to write `MSSQLExploiter` after all.